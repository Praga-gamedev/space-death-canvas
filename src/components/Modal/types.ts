export interface IModalProps {
    show: boolean;
    onClose: () => void;
}
