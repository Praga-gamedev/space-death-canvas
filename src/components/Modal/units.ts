import styled from '@emotion/styled';

export const ModalOverlay = styled.div`
    display: ${({ show }: { show: boolean }) => (show ? 'flex' : 'none')};

    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    background-color: rgba(0, 0, 0, 0.7);
    z-index: 2000;
`;

export const ModalBody = styled.div`
    background-color: #fff;
    padding: 20px;
    border-radius: 15px;
    cursor: auto;
    z-index: 2001;
`;
