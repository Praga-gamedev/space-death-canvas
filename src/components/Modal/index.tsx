import React, { FC, memo, useRef } from 'react';

import { S } from './units';
import { IModalProps } from './types';

export const Modal: FC<IModalProps> = memo(({ show, onClose, children }) => {
    const overlayRef = useRef();

    const close = (event: Event) => {
        if (event.target === overlayRef.current) {
            onClose();
        }
    };

    return (
        <S.ModalOverlay show={show} ref={overlayRef} onClick={close}>
            <S.ModalBody>{children}</S.ModalBody>
        </S.ModalOverlay>
    );
});
