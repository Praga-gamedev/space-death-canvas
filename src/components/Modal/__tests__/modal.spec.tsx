import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { Modal } from '@components/Modal';

const defaultModalProps = { show: true, onClose: () => {} };

describe('Modal component', () => {
    test('modal rendered correctly', () => {
        const ModalComponent = renderer
            .create(<Modal {...defaultModalProps} />)
            .toJSON();

        expect(ModalComponent).toMatchSnapshot();
    });

    test('modal content rendered correctly', () => {
        const modalBody = 'modal text';
        const modalContentId = 'modalContent';

        const ModalComponent = mount(
            <Modal {...defaultModalProps}>
                <div data-test-id={modalContentId}>{modalBody}</div>
            </Modal>
        );
        const ModalContent = ModalComponent.find({
            'data-test-id': modalContentId,
        });

        expect(ModalContent).toBeDefined();
        expect(ModalContent.text()).toBe(modalBody);
    });

    test('close event working', () => {
        const onClose = jest.fn();
        const props = { onClose, show: true };

        const ModalComponent = mount(<Modal {...props} />);
        ModalComponent.find({ 'data-test-id': 'modalOverlay' })
            .hostNodes()
            .simulate('click');

        expect(onClose).toBeCalledTimes(1);
    });

    test('close event not triggered if click on body', () => {
        const onClose = jest.fn();
        const modalBody = 'Body';
        const props = { onClose, show: true, children: modalBody };

        const ModalComponent = mount(<Modal {...props} />);
        ModalComponent.find({ children: modalBody })
            .hostNodes()
            .simulate('click');

        expect(onClose).not.toHaveBeenCalled();
    });
});
