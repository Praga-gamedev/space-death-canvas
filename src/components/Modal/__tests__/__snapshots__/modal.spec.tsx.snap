// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component modal rendered correctly 1`] = `
<div
  className="css-17k3zxr"
  data-test-id="modalOverlay"
  onClick={[Function]}
>
  <div
    className="css-n40hnu"
  />
</div>
`;
